reader.go

```diff
@@ -827,6 +858,9 @@ func (r *Reader) handshake() error {
 		_ = conn.Close()
 	}()
 
+	// wait for topic to exist
```

Comment:
As mentioned in #200, I don't think this is a good place to insert the wait as it will cause the rebalance to time out if the topic doesn't exist.  I think that as you test it, you should be able to produce that condition fairly easily, but LMK if you find otherwise!

Reply:
Hrm, I'm confused. This is being called before rebalance, and therefore before joinGroup and syncGroup. Also, joinGroup is the only place I see config.RebalanceTimeout being used. There's no rebalance/join to timeout at this point, unless I'm missing something.

Reply:
This is a little bit subtle.  In a nutshell, there are three types of participants in consumer groups: the coordinator, the leader (which is a special consumer), and the consumers.  The coordinator is a Kafka broker whose responsibility is to manage group membership and coordinate rebalances.

When a rebalance occurs:
1. All the consumers phone in to the coordinator to join the group: https://github.com/segmentio/kafka-go/blob/d19f52f08673027cb346c63eb317b8e644b101af/reader.go#L303
1. Via the join group response, one of the consumers discovers that it has been appointed the leader and will be responsible for generating assignments: https://github.com/segmentio/kafka-go/blob/d19f52f08673027cb346c63eb317b8e644b101af/reader.go#L337
1. Upon successful join, all consumers *including the leader* call sync group. https://github.com/segmentio/kafka-go/blob/d19f52f08673027cb346c63eb317b8e644b101af/reader.go#L410  The leader will produce assignments and include them in the sync group request.  The other consumers leave that part of the request empty. https://github.com/segmentio/kafka-go/blob/d19f52f08673027cb346c63eb317b8e644b101af/reader.go#L369
1. The coordinator takes the leader's assignments and returns them to the other consumers via the sync group response.

So if you put in a blocking call before generating assignments, your leader won't ever send the sync group request with the assignments.  If that happens, the coordinator will eventually time out the sync, evict the leader from the group, and then run another rebalance.  It's not immediately obvious from reading the code that all this is happening, but you can read up on the protocol here: https://cwiki.apache.org/confluence/display/KAFKA/Kafka+Client-side+Assignment+Proposal.

As with the case of more consumers than partitions, it's not a protocol level error to request a non-existent topic...it's a perfectly valid state.  It's kind of odd to me that the CG protocol doesn't provide a way to flag that as an error, but that's not for us mere mortals to debate. 😉   Anyway, that's why I was recommending to return empty assignments and then leverage the topic watcher feature that's already built in to trigger the rebalance when the topic appears.  Does that make sense?

Reply:
@stevevls What about moving the wait for topic logic earlier? 
Maybe inside of the run func? 
Like here https://github.com/segmentio/kafka-go/blob/master/reader.go#L867

From my understanding there isn't any reason to move into the handshake logic if the topic doesn't exist. By moving it higher we can also reduce the amount of times we can block and reduce the overall amount of network calls made.

Reply:
@ShaneSaww That's an interesting idea, though I wonder how it would handle the case where you have a large group of consumers all waiting for a topic to exist.  The thing is, their wait times won't be synchronized, so you will have an initial flurry of rebalances as the consumers discover the existence of the topic and join the group.

Contrast that to where the group has already gone through the handshake.  In that case, only a single rebalance would be triggered.  Furthermore, having gone through handshake, all the consumers will be in sync and heartebeating, so the first member to detect the existence of the topic will force the rebalance.

What do you think?

Reply:
I believe the broker will wait a few seconds for new consumers to show up before rebalancing, controlled by `group.initial.rebalance.delay.ms` which defaults to 3 seconds. So if the retry loop is short enough, you should only get one or two rebalances, right?

Reply:
I think the number of rebalances would depend on a couple of things.  First would be the setting for `group.initial.rebalance.delay.ms`, second would be the sleep time between polling to check the topic, and third would be the number of consumers in the group.  Depending on those three inputs, you easily have more than 2 rebalances...especially if the number of consumers is large and there's a big difference between the rebalance delay and the sleep on the topic loop.  Those timeouts are configurable (or should be if they're not), so we should be careful about making assumptions about their values.

The more we talk this out, the more convinced I am that we the only way to _guarantee_ a single rebalance is to succeed the initial consumer group rebalance when the topic doesn't exist and then trigger a rebalance when it comes into existence.

As a side note, kafka-go doesn't currently support the `Reader` using consumer groups to pull messages from multiple topics.  However, the Kafka protocol allows for it.  If we were to implement that feature later, I could imagine that if we ended up in case where a group asks to consume two topics and only one exists, we wouldn't want to block consuming that one until the other one comes into existence.  I think that's another reason it would be preferable to create a consumer group for the non-existent topic.

I realize that all said and done, it's a weird case to be trying to consume a topic that doesn't exist, and this is a lot of thought going into what's an edge case.  But I think that it's best to handle edge cases as gracefully as possible, so I'd propose we go with what I suggested.  If you have strong objections, then let's keep talking through them!

Reply:
@stevevls I would deal with that by having the wait time between calls for checking the topic to be fixed (and with what @abuchanan-nr  said ideally less then `group.initial.rebalance.delay.ms`).  That way the rebalance jitter _should_ be similar to just starting all the consumers without the topic checking. 
I don't think having them all ask for the same topic info is too hard on the brokers.  

The other way makes sense to me as well, it feels a little strange as the coordinator will technically get an error when trying to  get the topic info to build the assignments.

Reply:
@stevevls  I didn't even think about the double topic case. That's an interesting one, and I totally see the validity in moving the check into the partitionWatcher in that case. 

Add adding all that config + trying to get the numbers correct to minimize rebalances makes the API quite confusing.

Reply:
@ShaneSaww @abuchanan-nr I started to get curious about how other libraries handle the case of unknown topics, so I looked at Sarama and librdkafka.

Sarama handles it like kafka-go currently does--if the topic doesn't exist, the CG leader throws an error and the rebalance goes into a failure loop.  Not terribly desirable.  

AFAICT, librdkafka ignores unknown topics at assignment time: https://github.com/edenhill/librdkafka/blob/5140cd9d965406a144de067b775bc9ca5255aea1/src/rdkafka_cgrp.c#L924.  But it does succeed the rebalance.  I would appreciate another set of eyes on that code b/c I'm not a C++ or librdkafka expert. 😄 Either way, I think we should mimic whatever librdkafka is doing.

On another note, I think we've gotten into the weeds a bit with handling the unknown topic.  One thing we can consider is merging in the fix for number consumers > number of partitions and then solve the unknown topic case in a separate PR.  Up to you!

Reply:
@stevevls I can look into the java code as well.

Reply:
Thanks for all the great discussion.

I think it's important to move forward here. There are some important cases mentioned in #200 that could be very common. 

We can revisit to optimize rebalances for unknown topics in future work. I agree it's important, but might take awhile.

We can either use this PR as-is, or I could move the `waitForTopic` into `run` where it would have its own connection that is not associated with a group coordinator.

Thoughts?

Reply:
@ShaneSaww Did you get a chance to see how the java client works?  I'm still of the opinion that we should take the librdkafka approach where a non-existent topic simply results in empty assignments, but I think it's most important to behave like other official libraries in case there are places where folks need to inter-operate between languages.

Reply:
Yeah, from what I can tell with the java client it does something like 
- start everything as normal
- wait for metadata refresh 
- if metdata refresh shows a topic that you care about rebalance